src/ws/session-utils.c

```diff
@@ -298,8 +299,13 @@ open_session (pam_handle_t *pamh)
 
       debug ("opening pam session for %s", name);
 
+      res = snprintf (home_env_buf, sizeof (home_env_buf), "HOME=%s", pwd->pw_dir);
```

Comment:
Why do we use stack allocated arrays at all. This seems risky in a setuid program. What about asprintf()?

Reply:
It's not stack allocated, it's a `static char`, and thus in the .data segment. It's the most "safe" allocation I'm aware of, and as it already uses the same approach for `pwd_string_buf`, it feels right to do the same for this string. It's also both cheaper and safer than heap allocation (asprintf), as that can always potentially fail.